### PR TITLE
HADOOP-19219. Add JPMS options required by hadoop-common

### PR DESCRIPTION
**Pull request by pan3793**
### Description of PR

JIRA: HADOOP-19219. Add JPMS options required by hadoop-common.

Adding JPMS ([JEP 261: Module System](https://openjdk.org/jeps/261)) options that required by JDK 17+ for testing (maven-surefire-plugin) and runtime (`hadoop-functions.sh`)

This PR is based on https://github.com/apache/hadoop/pull/6939.

### How was this patch tested?

Tested on x86 Linux Java 17, this PR fixes the following cases:

hadoop-auth
```
[ERROR] org.apache.hadoop.security.authentication.server.TestLdapAuthenticationHandler  Time elapsed: 1.14 s  <<< ERROR!
java.lang.IllegalAccessError: class org.apache.directory.server.core.security.CertificateUtil (in unnamed module @0x3196eefd) cannot access class sun.security.x509.X500Name (in module java.base) because module java.base does not export sun.security.x509 to unnamed module @0x3196eefd
[ERROR] org.apache.hadoop.security.authentication.server.TestMultiSchemeAuthenticationHandler  Time elapsed: 1.299 s  <<< ERROR!
java.lang.IllegalAccessError: class org.apache.directory.server.core.security.CertificateUtil (in unnamed module @0x54a097cc) cannot access class sun.security.util.ObjectIdentifier (in module java.base) because module java.base does not export sun.security.util to unnamed module @0x54a097cc
```

hadoop-common
```
[ERROR] Tests run: 9, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.155 s <<< FAILURE! - in org.apache.hadoop.io.compress.TestCodecPool
[ERROR] testCompressorConf(org.apache.hadoop.io.compress.TestCodecPool)  Time elapsed: 0.008 s  <<< ERROR!
java.lang.reflect.InaccessibleObjectException: Unable to make field private int java.util.zip.Deflater.level accessible: module java.base does not "opens java.util.zip" to unnamed module @7ca48474
```

### For code changes:

- [x] Does the title or this PR starts with the corresponding JIRA issue id (e.g. 'HADOOP-17799. Your PR title ...')?
- [ ] Object storage: have the integration tests been executed and the endpoint declared according to the connector-specific documentation?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the `LICENSE`, `LICENSE-binary`, `NOTICE-binary` files?

### CODE REVIEW

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |  12m  0s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  0s |  |  detect-secrets was not available.  |
| +0 :ok: |  shelldocs  |   0m  0s |  |  Shelldocs was not available.  |
| +0 :ok: |  xmllint  |   0m  0s |  |  xmllint was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| -1 :x: |  test4tests  |   0m  0s |  |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
|||| _ trunk Compile Tests _ |
| +0 :ok: |  mvndep  |  15m  6s |  |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  32m 59s |  |  trunk passed  |
| +1 :green_heart: |  compile  |  17m 24s |  |  trunk passed with JDK Ubuntu-11.0.24+8-post-Ubuntu-1ubuntu320.04  |
| +1 :green_heart: |  compile  |  16m 33s |  |  trunk passed with JDK Private Build-1.8.0_422-8u422-b05-1~20.04-b05  |
| +1 :green_heart: |  mvnsite  |   2m 26s |  |  trunk passed  |
| +1 :green_heart: |  javadoc  |   1m 59s |  |  trunk passed with JDK Ubuntu-11.0.24+8-post-Ubuntu-1ubuntu320.04  |
| +1 :green_heart: |  javadoc  |   1m 35s |  |  trunk passed with JDK Private Build-1.8.0_422-8u422-b05-1~20.04-b05  |
| +1 :green_heart: |  shadedclient  |  38m 21s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 33s |  |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   1m 14s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  17m 18s |  |  the patch passed with JDK Ubuntu-11.0.24+8-post-Ubuntu-1ubuntu320.04  |
| +1 :green_heart: |  javac  |  17m 17s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  16m 14s |  |  the patch passed with JDK Private Build-1.8.0_422-8u422-b05-1~20.04-b05  |
| +1 :green_heart: |  javac  |  16m 14s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| +1 :green_heart: |  mvnsite  |   2m 23s |  |  the patch passed  |
| +1 :green_heart: |  shellcheck  |   0m  4s |  |  No new issues.  |
| +1 :green_heart: |  javadoc  |   1m 53s |  |  the patch passed with JDK Ubuntu-11.0.24+8-post-Ubuntu-1ubuntu320.04  |
| +1 :green_heart: |  javadoc  |   1m 34s |  |  the patch passed with JDK Private Build-1.8.0_422-8u422-b05-1~20.04-b05  |
| +1 :green_heart: |  shadedclient  |  40m 40s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   0m 38s |  |  hadoop-project in the patch passed.  |
| +1 :green_heart: |  unit  |  20m 24s |  |  hadoop-common in the patch passed.  |
| +1 :green_heart: |  asflicense  |   1m  5s |  |  The patch does not generate ASF License warnings.  |
|  |   | 238m 34s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.47 ServerAPI=1.47 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7084/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/7084 |
| Optional Tests | dupname asflicense mvnsite unit codespell detsecrets shellcheck shelldocs compile javac javadoc mvninstall shadedclient xmllint |
| uname | Linux c628261a26ac 5.15.0-117-generic #127-Ubuntu SMP Fri Jul 5 20:13:28 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / 5176c4e8566524e99db8cbfeccf42e20e0138e56 |
| Default Java | Private Build-1.8.0_422-8u422-b05-1~20.04-b05 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.24+8-post-Ubuntu-1ubuntu320.04 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_422-8u422-b05-1~20.04-b05 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7084/1/testReport/ |
| Max. process+thread count | 3152 (vs. ulimit of 5500) |
| modules | C: hadoop-project hadoop-common-project/hadoop-common U: . |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7084/1/console |
| versions | git=2.25.1 maven=3.6.3 shellcheck=0.7.0 |
| Powered by | Apache Yetus 0.14.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by slfan1989** [comment]
LGTM.

File: hadoop-common-project/hadoop-common/src/main/bin/hadoop-functions.sh [hunk 1]
```diff
@@ -1569,6 +1569,19 @@ function hadoop_finalize_hadoop_opts
...
 
+## @description  Finish configuring JPMS that enforced for JDK 17 and higher
+## @description  prior to executing Java
+## @audience     private
+## @stability    evolving
+## @replaceable  yes
+function hadoop_finalize_jpms_opts
+{
+    hadoop_add_param HADOOP_OPTS IgnoreUnrecognizedVMOptions "-XX:+IgnoreUnrecognizedVMOptions"
+    hadoop_add_param HADOOP_OPTS open.sun.security.x509 "--add-opens=java.base/java.util.zip=ALL-UNNAMED"
```
Comment (slfan1989) [inline hadoop-common-project/hadoop-common/src/main/bin/hadoop-functions.sh]: Should it be `java.util.zip`?

Comment (pan3793) [reply hadoop-common-project/hadoop-common/src/main/bin/hadoop-functions.sh]: oh, right. it's a copy-paste issue, fixed.

**Review by slfan1989 — APPROVED** [review]

File: hadoop-project/pom.xml [hunk 2]
```diff
@@ -167,8 +167,14 @@
     <enforced.java.version>[${javac.version},)</enforced.java.version>
     <enforced.maven.version>[3.3.0,)</enforced.maven.version>
 
+    <extraJavaTestArgs>
```
Comment (slfan1989) [inline hadoop-project/pom.xml:170]: I haven't decided on the name for this tag yet, but I feel that calling it `extraJavaTestArgs` may not be the best option.

Comment (pan3793) [reply hadoop-project/pom.xml:170]: the name is borrowed from the Spark project.

Comment (slfan1989) [reply hadoop-project/pom.xml:170]: This is not a blocking concern, as it relates to JDK-specific parameters. I'm unsure whether it should be named something related to the JDK. I have no objections if we keep it as is.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 31s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  0s |  |  detect-secrets was not available.  |
| +0 :ok: |  shelldocs  |   0m  0s |  |  Shelldocs was not available.  |
| +0 :ok: |  xmllint  |   0m  0s |  |  xmllint was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| -1 :x: |  test4tests  |   0m  0s |  |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
|||| _ trunk Compile Tests _ |
| +0 :ok: |  mvndep  |  15m 38s |  |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  39m 19s |  |  trunk passed  |
| +1 :green_heart: |  compile  |  19m 47s |  |  trunk passed with JDK Ubuntu-11.0.24+8-post-Ubuntu-1ubuntu320.04  |
| +1 :green_heart: |  compile  |  16m 13s |  |  trunk passed with JDK Private Build-1.8.0_422-8u422-b05-1~20.04-b05  |
| +1 :green_heart: |  mvnsite  |   2m 23s |  |  trunk passed  |
| +1 :green_heart: |  javadoc  |   1m 54s |  |  trunk passed with JDK Ubuntu-11.0.24+8-post-Ubuntu-1ubuntu320.04  |
| +1 :green_heart: |  javadoc  |   1m 30s |  |  trunk passed with JDK Private Build-1.8.0_422-8u422-b05-1~20.04-b05  |
| +1 :green_heart: |  shadedclient  |  40m 52s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 33s |  |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   1m 14s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  18m 23s |  |  the patch passed with JDK Ubuntu-11.0.24+8-post-Ubuntu-1ubuntu320.04  |
| +1 :green_heart: |  javac  |  18m 23s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  18m  7s |  |  the patch passed with JDK Private Build-1.8.0_422-8u422-b05-1~20.04-b05  |
| +1 :green_heart: |  javac  |  18m  7s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| +1 :green_heart: |  mvnsite  |   2m 23s |  |  the patch passed  |
| +1 :green_heart: |  shellcheck  |   0m  5s |  |  No new issues.  |
| +1 :green_heart: |  javadoc  |   1m 42s |  |  the patch passed with JDK Ubuntu-11.0.24+8-post-Ubuntu-1ubuntu320.04  |
| +1 :green_heart: |  javadoc  |   1m 25s |  |  the patch passed with JDK Private Build-1.8.0_422-8u422-b05-1~20.04-b05  |
| +1 :green_heart: |  shadedclient  |  46m 51s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   0m 32s |  |  hadoop-project in the patch passed.  |
| +1 :green_heart: |  unit  |  20m 52s |  |  hadoop-common in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 57s |  |  The patch does not generate ASF License warnings.  |
|  |   | 247m 51s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.47 ServerAPI=1.47 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7084/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/7084 |
| Optional Tests | dupname asflicense mvnsite unit codespell detsecrets shellcheck shelldocs compile javac javadoc mvninstall shadedclient xmllint |
| uname | Linux 19875fac907b 5.15.0-117-generic #127-Ubuntu SMP Fri Jul 5 20:13:28 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / 0dc42895f1642d3b7ad54dba95e62b46f13af6dc |
| Default Java | Private Build-1.8.0_422-8u422-b05-1~20.04-b05 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.24+8-post-Ubuntu-1ubuntu320.04 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_422-8u422-b05-1~20.04-b05 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7084/2/testReport/ |
| Max. process+thread count | 3152 (vs. ulimit of 5500) |
| modules | C: hadoop-project hadoop-common-project/hadoop-common U: . |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7084/2/console |
| versions | git=2.25.1 maven=3.6.3 shellcheck=0.7.0 |
| Powered by | Apache Yetus 0.14.0 https://yetus.apache.org |


This message was automatically generated.

**Review by steveloughran — APPROVED** [review]

**Comment by slfan1989** [comment]
@pan3793 Thanks for the contribution! I want to push directly to branch-3.4, but if this PR involves compilation, it's better to submit a separate PR to handle it. If you have time, please also submit a PR for branch-3.4. @steveloughran Thanks for the review!

**Pull request merged**